### ISSUE TITLE
The results in the mnist example are wrong/misleading

### ISSUE THREAD

**Issue by themasterlink** [Rishit-dagli/Gradient-Centralization-TensorFlow#24]
**Describe the bug**
The results in your colab ipython notebook are misleading: https://colab.research.google.com/github/Rishit-dagli/Gradient-Centralization-TensorFlow/blob/main/examples/gctf_mnist.ipynb

In this example, the model is first trained with a normal Adam optimizer: 

```python
model.compile(optimizer = tf.keras.optimizers.Adam(),
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])

history_no_gctf = model.fit(training_images, training_labels, epochs=5, callbacks = [time_callback_no_gctf])
```

And afterwards the same model is recompiled with the `gctf.optimizers.adam()`. However, recompiling a keras model does not reset the weights. This means that in the first `fit` call the model is trained and then in the second `fit` call with the new optimizer the same model is used and of course then the results are better. 

This can be fixed, by recreating the model for the second run, by just adding these few lines:

```python
import gctf #import gctf

time_callback_gctf = TimeHistory()

# Model architecture
model = tf.keras.models.Sequential([
                                    tf.keras.layers.Flatten(), 
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(256, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(64, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(256, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(64, activation=tf.nn.relu), 
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer = gctf.optimizers.adam(),
              loss = 'sparse_categorical_crossentropy',
              metrics=['accuracy'])

history_gctf = model.fit(training_images, training_labels, epochs=5, callbacks=[time_callback_gctf])
```

However, then the results are not better than without `gctf`:
```
Type                   Execution time    Accuracy      Loss
-------------------  ----------------  ----------  --------
Model without gctf:           24.7659    0.88825   0.305801
Model with gctf               24.7881    0.889567  0.30812
``` 

Could you please clarify what happens here. I tried this `gctf.optimizers.adam()` optimizer in my own research and it didn't change the results at all and now after seeing it doesn't work in the example which was constructed here. Makes me question the results of this paper.

**To Reproduce**
Execute the colab file given in the repository: https://colab.research.google.com/github/Rishit-dagli/Gradient-Centralization-TensorFlow/blob/main/examples/gctf_mnist.ipynb

**Expected behavior**
The right comparison would be if both models start from a random initialization, not that the second model can start with the already pre-trained weights.

Looking forward to a fast a swift explanation.

Best,
Max

**Comment by themasterlink** [comment]
**Addition**
This is also true for the `horses-v-humans` example. The crucial bug here is that `model.compile` does not reset any learned weights:

https://stackoverflow.com/a/47996024

**Comment by Rishit-dagli** [comment]
Hi,

I tried reproducing with the code edits you asked for with TensorFlow Enterprise 2.5 with an NVIDIA Tesla V100 on a GCP AI Platform Instance using an M71 VM Image and as you can see I am not able to reproduce these. Find the gist [here](https://gist.github.com/Rishit-dagli/e07b302be9594c8aaf31ed0ee285d5d8). Following this, I am closing this issue.

Also, you mentioned using this in your research, the authors did mention using this only for limited kinds of problems.
